datalad/cmd.py

```diff
@@ -6,66 +6,152 @@
 #   copyright and license terms.
 #
 # ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""
+Wrapper for command and function calls, allowing for dry runs and output handling
+
+"""
 
-import os
-import commands
 
+import subprocess
+import sys
 import logging
-lgr = logging.getLogger('datalad.files')
+import os
+
+lgr = logging.getLogger('datalad.cmd')
 
-def dry(s, dry_):
-    """Small helper for dry runs
-    """
-    if dry_:
-        return "DRY " + s
-    return s
 
 class Runner(object):
-    """Helper to run commands and functions while allowing for dry_run and later reporting
+    """Provides a wrapper for calling functions and commands.
+
+    An object of this class provides a methods calls shell commands or python functions,
+    allowing for dry runs and output handling.
+
+    Outputs (stdout and stderr) can be either logged or streamed to system's stdout/stderr during execution.
+    This can be enabled or disabled for both of them independently.
+    Additionally allows for dry runs. This is achieved by initializing the `Runner` with `dry=True`.
+    The Runner will then collect all calls as strings in `commands`.
     """
 
     __slots__ = ['commands', 'dry']
 
     def __init__(self, dry=False):
+        self.dry = dry
         self.commands = []
-        self.dry = dry                    # TODO: make use of it
 
-    # TODO -- remove dry_run specification here -- use constructor parameter
-    #def __call__(self, cmd, dry_run=False):
-    def getstatusoutput(self, cmd, dry_run=None):
-        """A wrapper around commands.getstatusoutput
+    def __call__(self, cmd, *args, **kwargs):
+        """Convenience method
+
+        This will call run() or call() depending on the kind of `cmd`.
+        If `cmd` is a string it is interpreted as the to be executed command.
+        Otherwise it is expected to be a callable.
+        Any other argument is passed to the respective method.
+
+        Parameters
+        ----------
+        cmd: str or callable
+           command string to be executed via shell or callable to be called.
+
+        *args and **kwargs:
+           see Runner.run() and Runner.call() for available arguments.
 
-        Provides improved logging for debugging purposes and raises
-        RuntimeError exception in case of non-0 return
+        Raises
+        ------
+        ValueError if cmd is neither a string nor a callable.
         """
+
+        if isinstance(cmd, basestring):
+            self.run(cmd, *args, **kwargs)
+        elif callable(cmd):
+            self.call(cmd, *args, **kwargs)
+        else:
+            raise ValueError("Argument 'command' is neither a string nor a callable.")
+
+    def run(self, cmd, log_stdout=False, log_stderr=True):
```

Comment:
please lets make it log_stdout=True by default to maintain close to previous logging behaviour.. while running "interactive" commands (e.g. add), just explicitly state to not  log